src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -340,6 +340,11 @@ public DiagnosticQueue DiagnosticQueue
             }
         }
 
+        /// <summary>
+        /// Callback to add additional files to be generated to disk.  Null if generating files is not desired.
+        /// </summary>
+        public Action<(string filePath, SourceText text)>? AddOutputFile;
```

Comment:
So one thing to consider for the LSIF case is the files are _huge_, they can easily be hundreds of megabytes in size. The LSIF tool today opens a stream and starts writing, and can often flush things while it's analyzing each document as things go. I'm not exactly sure what the API would look like, but it might be necessary to have an API which is "CreateOutputFile" which instead returns a Stream the analyzer can just party on directly. The shortcut of "just add a string/source text" could of course still be helper methods to optimize the common case.

Reply:
how huge is huge?  per file are we talkign MBs, 10s of MBs?  100s?

Reply:
@CyrusNajmabadi Hundreds.

Reply:
(Or at least many tens.)

Reply:
hrmm... let me take offline with you.

Reply:
ok.  new approach here creates an actual stream as hte lowest level abstraction an ArtifactGenerator can write to.   there are also helpers for strings, source-texts, and stringbuilders.